File: clients/desktop/GraNAS.Desktop.App/Services/Api/PermissionsApi.cs
using GraNAS.Desktop.Contracts.Metadata;

namespace GraNAS.Desktop.App.Services.Api;

public class PermissionsApi : ApiBase, IPermissionsApi
{
  public PermissionsApi(HttpClient http) : base(http) { }

  public Task<List<PermissionResponse>> GetPermissionsAsync(Guid folderId, CancellationToken ct = default)
    => GetAsync<List<PermissionResponse>>($"api/metadata/folders/{folderId}/permissions", ct);

  public Task<PermissionResponse> GrantAsync(Guid folderId, GrantPermissionRequest request, CancellationToken ct = default)
    => PostAsync<PermissionResponse>($"api/metadata/folders/{folderId}/permissions", request, ct);

  public Task RevokeAsync(Guid folderId, Guid userId, CancellationToken ct = default)
    => DeleteAsync($"api/metadata/folders/{folderId}/permissions/{userId}", ct);
}

File: clients/web/src/api/permissions.api.ts
import type { GrantPermissionRequest, PermissionResponse } from '../types/permission';
import { api } from './client';

export const permissionsApi = {
  grant: (folderId: string, data: GrantPermissionRequest) =>
    api.post<PermissionResponse>(`/api/metadata/folders/${folderId}/permissions`, data),

  revoke: (folderId: string, userId: string) =>
    api.delete(`/api/metadata/folders/${folderId}/permissions/${userId}`),
};
